### DISCUSSION TITLE
Proper way to test using ebean-test when class has dependencies

### DISCUSSION THREAD

**Discussion by Brian-Hofmann** [ebean-orm/ebean#3244] [General]
I'm starting to go through my spring service and add tests. I have read the ebean docs referencing the new way to do testing via Docker containers. I was able to get some tests written for my Ebean classes and can confirm that it properly runs on the docker container.

However, I am trying to write tests for a @Service within my Spring app and am trying to get it to run the tests on the docker container, but it is running them against my local database instead. I believe this may have to do with how Spring auto wire works alongside @SpringBootTest

Example:

```Java
@SpringBootTest
public class MyServiceTest {

    @Autowired
    private MyService myService;

    @Test
    public void myTest() {
       // Do some initial DB persistence for test
       MyResult result = myService.someFunction();
       // Assertions
   }

}
```

When I run this test individually not through the mvn run tests command, the test runs fine and does not add entries into my local database. Is there a way to configure the mvn command to point to the spun up docker container?

Also side question. Whenever I use UserContext within this test, the insertions fail saying the created_by cannot be null. Is there a way you recommend setting up dependency tests like this so that Spring autowire does not interfere?

Thanks in advance!

**Comment by rob-bygrave** [comment]
The above is not clear in that there is no reference to ebean Database or DB in the above.  

Maybe git clone the Spring boot example project and submit a PR showing the issue?  https://github.com/ebean-orm-examples/example-springboot/blob/master/src/test/resources/application-test.properties

**Reply by Brian-Hofmann** [reply]
My apologies. I should have been more clear on where ebean comes in.

In the service, there are calls to a repository which makes use of the Ebean DB class. 

The service looks similar to this

```java
@Service
public class MyService {

   @Autowired
   MyEntityRepository myEntityRepository;

   public someFunction() {
      MyEntity myEntity = new MyEntity();
      myEntityRepository.save(myEntity);
  }
```

And the repository like so

```java
@Repository
public class MyEntityRepository {

   public save(MyEntity entityToSave) {
      DB.save(entityToSave)
  }
```

The database is configured in a config class where the values are populated from the appropriate yml file. 

```java
@Configuration
public class DatabaseConfiguration {
    @Value("${db-url}")
    private String dbUrl;

    @Value("${db-username}")
    private String dbUsername;

    @Value("${db-password}")
    private String dbPassword;

    @Value("${db-driver}")
    private String dbDriver;

    @Value("${ebean.migration.run}")
    private boolean autoRunMigration;

    @Bean
    public Database getDataSource() {
        DataSourceConfig dsConfig = new DataSourceConfig();
        dsConfig.setUsername(dbUsername);
        dsConfig.setPassword(dbPassword);
        dsConfig.setUrl(dbUrl);
        dsConfig.setDriver(dbDriver);

        DatabaseConfig dbConfig = new DatabaseConfig();
        dbConfig.setDataSourceConfig(dsConfig);
        dbConfig.loadFromProperties();
        dbConfig.setRunMigration(autoRunMigration);
        dbConfig.setTenantMode(TenantMode.PARTITION);
        dbConfig.setTenantPartitionColumn("tenant_id");
        dbConfig.setCurrentTenantProvider(new MyTenantProvider());
        dbConfig.setCurrentUserProvider(new MyCurrentUserProvider());

        return DatabaseFactory.create(dbConfig);
    }
```

I have the values for application-test.yml

```yml
ebean:
  migration:
    run: true
  test:
    platform: postgres
    ddlMode: dropCreate
    dbName: app_testing
```

I have a bootstrap.yml file for local testing and it seems like Spring is reading that file and persisting to the DB set there instead of the docker one only when I run the command mvn test instead of running the test as a one off. 

bootstrap.yml file:

```yml
db-url: jdbc:postgresql://localhost:5432/myDB
db-username: username
db-password: password
db-driver: org.postgresql.Driver
ebean:
  migration:
    run: true
```

I can confirm that the test without the service dependency runs fine and points to the docker DB. 

That test just looks like this:

```Java
public class MyEntityTest {
    @Test
    public void saveEntity() {
      MyEntity entityToSave = new MyEntity(...params);
      DB.save(entityToSave);
   }
}
```

In the test with the dependency, I can also confirm that the actual implementation of my CurrentUserProvider is called and calls out to the Spring UserContext instead of the Ebean one which makes me think the dependency test causes Spring to load things differently perhaps.

Also it looks like even running the test as a one off, the UserContext is not properly being set when I set it with the ebean UserContext class at the start of the test. Saving errors with "cannot insert null into created_by column"

If you need more information, I can use the repo you provided and try to reproduce this on it

**Comment by Brian-Hofmann** [comment]
Yeah after some more investigation, this looks like a Database configuration issue on load. The test without the dependency is pointing to the correct docker database (it does not run the database config file), but the test with one is pointing to my local DB and reading like it is running the app (runs the database config file and has my current user and tenant providers). Is there a different way I need to configure the DB at first when the test has an autowired dependency?

**Reply by rbygrave** [reply]
Are you able to reproduce the issue with https://github.com/ebean-orm-examples/example-springboot ?  perhaps with some modifications to that example project?

**Reply by Brian-Hofmann** [reply]
I'll spin up an example and let you know!

**Reply by Brian-Hofmann** [reply]
I was able to reproduce the issue here: https://github.com/Brian-Hofmann/example-springboot/tree/Discussion-3244-Repro

**Comment by Brian-Hofmann** [comment]
Any updates on this?

**Comment by Brian-Hofmann** [comment]
Figured this one out. Assuming your test profile is "test", place

```Java
@Profile("!test")
```

above your database configuration class. That ensures the default profile is not loaded which takes precedence over the ebean docker container running
